storage/innobase/handler/handler0alter.cc

```diff
@@ -8433,6 +8433,178 @@ innobase_update_foreign_cache(
+@param[in]	trx	transaction
+@param[in]	table	old table
+@return		boolean flag
+@retval	true	on failure
+@retval false	on success */
+static
+bool
+innobase_change_fields_versioning_try(
+	const Alter_inplace_info* ha_alter_info,
+	const ha_innobase_inplace_ctx* ctx,
+	trx_t* trx,
+	const TABLE* table)
+{
+	DBUG_ENTER("innobase_change_fields_versioning_try");
+
+	DBUG_ASSERT(ha_alter_info);
+	DBUG_ASSERT(ctx);
+
+	if (!(ha_alter_info->handler_flags & ALTER_COLUMN_UNVERSIONED)){
+		DBUG_RETURN(false);
+	}
+
+	uint virtual_count = 0;
+
+	List_iterator_fast<Create_field> it(
+	    ha_alter_info->alter_info->create_list);
+
+	for (uint i = 0; i < table->s->fields; i++) {
+		const Field* field = table->field[i];
+
+		if (innobase_is_v_fld(field)) {
+			virtual_count++;
+			continue;
+		}
+
+		const Create_field* create_field = NULL;
+		while (const Create_field* cf = it++) {
+			if (cf->field == field) {
+				create_field = cf;
+				break;
+			}
+		}
+		it.rewind();
+		DBUG_ASSERT(create_field);
+
+		if (create_field->versioning
+		    == Column_definition::VERSIONING_NOT_SET) {
+			continue;
+		}
+
+		const dict_table_t* table = ctx->new_table;
+		const dict_col_t* col
+		    = dict_table_get_nth_col(table, i - virtual_count);
+
+		DBUG_ASSERT(!col->vers_sys_start() && !col->vers_sys_end());
+
+		ulint new_prtype
+		    = create_field->versioning
+			      == Column_definition::WITHOUT_VERSIONING
+			  ? col->prtype & ~DATA_VERSIONED
+			  : col->prtype | DATA_VERSIONED;
+
+		if (innobase_change_field_versioning_try(
+			trx, table->id, col->name(*table), new_prtype)) {
+			DBUG_RETURN(true);
+		}
+	}
+
+	DBUG_RETURN(false);
+}
+
+/** Changes WITH/WITHOUT SYSTEM VERSIONING for fields
+in the data dictionary cache.
+@param ha_alter_info Data used during in-place alter
+@param ctx In-place ALTER TABLE context
+@param table MySQL table as it is before the ALTER operation */
+static
+void
+innobase_change_fields_versioning_cache(
+	Alter_inplace_info*		ha_alter_info,
+	const ha_innobase_inplace_ctx*	ctx,
+	const TABLE*			table)
+{
+	DBUG_ENTER("innobase_change_fields_versioning");
+
+	DBUG_ASSERT(ha_alter_info);
+	DBUG_ASSERT(ctx);
+	DBUG_ASSERT(ha_alter_info->handler_flags & ALTER_COLUMN_UNVERSIONED);
+
+	uint virtual_count = 0;
+
+	List_iterator_fast<Create_field> it(
+	    ha_alter_info->alter_info->create_list);
+
+	for (uint i = 0; i < table->s->fields; i++) {
+		const Field* field = table->field[i];
+
+		if (innobase_is_v_fld(field)) {
+			virtual_count++;
+			continue;
+		}
+
+		const Create_field* create_field = NULL;
+		while (const Create_field* cf = it++) {
+			if (cf->field == field) {
+				create_field = cf;
+				break;
+			}
+		}
+		it.rewind();
+		DBUG_ASSERT(create_field);
+
+		dict_col_t* col
+		    = dict_table_get_nth_col(ctx->new_table, i - virtual_count);
+
+		if (create_field->versioning
+		    == Column_definition::WITHOUT_VERSIONING) {
+			DBUG_ASSERT(!col->vers_sys_start()
+				    && !col->vers_sys_end());
```

Comment:
Never use `&&` at the top level in assertion expressions. Instead, split the assertion:
```c++
DBUG_ASSERT(!col->vers_sys_start());
DBUG_ASSERT(!col->vers_sys_end());
```